### PR TITLE
ref(auth): Refactor auth/helper.py for clarity

### PR DESCRIPTION
**Pull request by RyanSkonnord**
This is intended to be a refactor-in-place with no behavioral changes. The diff view may be noisy -- I recommend hiding whitespace changes, and it may help to review one commit at a time. The substance of the changes, summarized, is:

- I grouped several intimately related top-level functions in `helper.py` into the new `AuthIdentityHandler` class. Their common parameters (auth_provider, provider, organization, request) are the class's instance variables, which reduces the length of the functions' signatures. The module's "entry points" are public `AuthIdentityHandler` methods; helper methods are now marked with a leading underscore. My hope is that this reduces the overall cognitive weight of the `helper.py` module's list of responsibilities.
- I extracted new helper methods, mainly from `handle_attach_identity`, with the goal of reducing the vertical size of the function bodies. The main business value of these functions is in navigating the maze of if-else cases that make up the login flow. We probably will have to do some debugging/improvements to that flow in the near future, so the more digestible we can make the if-else structure, the better.
- Various drive-by opportunities to factor out duplicated code. I also wrapped the `if not auth.login` pattern into a try-except pattern that I believe makes the meaning a bit more obvious.

~~This introduces breaking changes for `getsentry` and requires https://github.com/getsentry/getsentry/pull/5744 when merged.~~ (No longer needed as of https://github.com/getsentry/sentry/pull/26571/commits/be32227476e22a6bbcf17525cef139a4ad6d9369. Should still complete https://github.com/getsentry/getsentry/pull/5744 after merging but it's not time-sensitive.)

From https://getsentry.atlassian.net/browse/ER-696

### CODE REVIEW

File: src/sentry/auth/helper.py
Comment (JoshFerge) [inline src/sentry/auth/helper.py:451]: this pattern of wrapping login with an exception feels weird to me -- would the ideal solution be to raise an exception within utils.login itself? i see that we don't really use the value returned from this function anywhere but here (guessing we rely on the side affects of login not completing to accomplish correct flow elsewhere).

it also seems like the only way login would return False is if the user has 2FA and hasn't completed it -- would `NotCompletedSecurityChecks` be a better fitting exception name?

File: src/sentry/auth/helper.py
Comment (JoshFerge) [inline src/sentry/auth/helper.py:277]: i kind of like checking `if auth_identity is None:` to keep similar as above and setting the `auth_is_new` variable within the block, this kind of confused me looking at it at first. your call though

**Review by JoshFerge — COMMENTED** [review]
looks solid to me, a few stylistic comments. nice refactor 👍🏼

**Review by JoshFerge — APPROVED** [review]
would be good to get 1 or 2 more approvals before merging, but lgtm

**Review by leedongwei — APPROVED** [review]
LGTM on shuffling things around

File: src/sentry/auth/helper.py
Comment (RyanSkonnord) [reply src/sentry/auth/helper.py:447]: > this pattern of wrapping login with an exception feels weird to me

It feels weird to me too in general, but as I understand it, the Pythonic style is more liberal about exceptions than most other languages. If you're trying to represent an exit condition from a function, an exception is preferred over a return value.

In this particular case, I do like it because most of the `if not auth.login(...):` blocks would interrupt the flow with a `return` statement. Putting flow control in the `except` block feels more natural to me than just any old code.

> would the ideal solution be to raise an exception within utils.login itself?

I would indeed love to make that change. I avoided making a change with such a large impact on other code, but...

> i see that we don't really use the value returned from this function anywhere but here (guessing we rely on the side affects of login not completing to accomplish correct flow elsewhere).

Interesting point. Let me take a look.

> it also seems like the only way login would return False is if the user has 2FA and hasn't completed it -- would `NotCompletedSecurityChecks` be a better fitting exception name?

Nice point. I'll double-check this and make the change.

File: src/sentry/auth/helper.py
Comment (RyanSkonnord) [reply src/sentry/auth/helper.py:273]: Exactly the kind of thing on which I want subjective feedback. 😉 Let's go with your way.

File: src/sentry/auth/helper.py
Comment (dcramer) [inline src/sentry/auth/helper.py:703]: wouldn't you want this to not create a new object on every access? e.g. should this be memoized?

**Review by dcramer — APPROVED** [review]
code change is too large for actual review - please test on staging before deploying to prod. also try to deploy to prod during low traffic period of day and make sure you're around to monitor.

File: src/sentry/auth/helper.py
Comment (RyanSkonnord) [reply src/sentry/auth/helper.py:703]: I believe it would create it only once in the lifespan of the typical `AuthHelper` object, but sure, I can make it to an instance variable.

Comment (dcramer) [reply src/sentry/auth/helper.py]: property just removes the need to put `()` on a function call, so everything time you do `self.auth_handler` that is the equiv of doing `self.auth_handler()` without the `@property` decorator. If you want it to lazy-instantiate, you can use `@cached_decorator` or `@memoize` (I forget which one we have)

Comment (RyanSkonnord) [reply src/sentry/auth/helper.py]: Understood. There's no need to lazy-instantiate. I was just extracting duplicated code and I think my Java instincts caused me to gravitate toward a method rather than an attribute. Either way, there are no performance implications -- `AuthIdentityHandler` is cheap to construct.

Comment (RyanSkonnord) [reply src/sentry/auth/helper.py:703]: Never mind, see https://github.com/getsentry/sentry/pull/26571/commits/3b3ae950ddf958b4d5da5c5380c453c26381709f. It does need to be instantiated upon use because `AuthHelper.auth_provider` is mutable.

**Comment by RyanSkonnord** [comment]
Development and testing complete. Planning to merge on Tuesday morning.

**Pull request merged**
